enhancements/api-review/add-ClusterImagePolicy-and-ImagePolicy-for-Signature-Verification.md

```diff
@@ -192,6 +197,36 @@ type FulcioCAWithRekor struct {
 	FulcioSubject PolicyFulcioSubject `json:"fulcioSubject,omitempty"`
 }
 
+// PKI defines the root of trust based on Root CA(s) and corresponding intermediate certificates.
+type PKI struct {
+	// CertificateAuthorityRootsData contains base64-encoded data of a certificate bundle PEM file, which contains one or more CA roots in the PEM format.
+	// +required
+	CertificateAuthorityRootsData []byte `json:"caRootsData,omitempty"`
+
+	// CertificateAuthorityIntermediatesData base64-encoded data of a certificate bundle PEM file, which contains one or more intermediate certificates in the PEM format.
+	// caIntermediatesData requires CertificateAuthorityRoots to be set.
+	// +optional
+	// +kubebuilder:validation:XValidation:rule="self == '' || self.CertificateAuthorityRootsData != ''", message="caIntermediatesData requires caRootsData to be set"
+	CertificateAuthorityIntermediatesData []byte `json:"caIntermediatesData,omitempty"`
+
+	// PKICertificateSubject must be set if any of the other fields are set.
+	// +required
+  // +kubebuilder:validation:XValidation:rule="self.CertificateAuthorityRootsData == '' || self.PKICertificateSubject != nil", message="pkiCertificateSubject must be set if caRootsData is specified"
+	PKICertificateSubject *PKICertificateSubject `json:"pkiCertificateSubject,omitempty"`
```

Comment:
how about just using `CertificateSubject` since we're already in the `PKI` struct?
```suggestion
	CertificateSubject *CertificateSubject `json:"certificateSubject,omitempty"`
```

Reply:
As far as correctness goes, sure, why not.

One concern to think about is that there are two audiences here:
- Users writing YAML: for them the `json` field names matter, and we can rely on the nesting structure to avoid repetition. So `certificateSubject` is better than `pkiCertificateSubject`
- Go programmers using the API package: all types are at the top level, so they should be unique standalone. So `PKICertificateSubject` is better than `CertificateSubject`, or maybe it should be `PKIImagePolicyCertificateSubject` in the extreme??
  - Note especially that the naming the top-level structure just `PKI` is not ideal… we can get away with that for `FulcioCAWithRekor`, but for `PKI` and the existing `PublicKey`… there’s some risk of clashes.

But, also, the YAML names are more important because they are user-facing; the Go type names are ~internal in a package with no stable API commitment (I think?) so if we ever encountered a conflict, we can rename things.

Reply:
```suggestion
	PKICertificateSubject *PKICertificateSubject `json:"certificateSubject,omitempty"`
```
Maybe only omit the PKI from the json annotation.

Reply:
> Note especially that the naming the top-level structure just PKI is not ideal… we can get away with that for FulcioCAWithRekor, but for PKI and the existing PublicKey… there’s some risk of clashes.

will `CustomPKI` instead of `PKI` make a difference? we can explain in the comment this means a PKI not from Sigstore.

Reply:
In the context of `openshift/api/config/v1alpha1` (or, worse, eventually, `…/v1`, I think `PKI` and `CustomPKI` are very similarly likely to clash — if there is going to be a clash, it’s going to be with some other non-`policy.json` PKI use.

See https://pkg.go.dev/github.com/openshift/api/config/v1 — there are things like `NamedCertificate` and `CertInfo`.

So I’d be tempted to use `ImagePolicy*` for all types (`ImagePolicyPKI`?), but that might well be an overreaction.

---

I do think the JSON/YAML field names are much more important; I’m not nearly as worried about the Go types.